### PR TITLE
Column auto generation

### PR DESCRIPTION
**Pull request by titovmx**

### CODE REVIEW

File: src/ng/components/column/column.list.js [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
+		const data = model.data;
+		const state = data();
+		
+		if (state.columns.length === 0 && this.generate) {
+			model.dataChanged.on(e => {
+				if (e.changes.hasOwnProperty('rows')) {
+					const deep = this.generateMode === 'deep' || isUndefined(this.generateMode);
+					if (!deep && this.generateMode !== 'flat') {
+						throw new AppError(
+							'template.core',
```
Comment (klumba12) [inline src/ng/components/column/column.list.js]: txception has invalid message and title

File: src/ng/components/column/column.list.js [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
 
 	onInit() {
+		const model = this.root.model;
+		const data = model.data;
+		const state = data();
+		
+		if (state.columns.length === 0 && this.generate) {
+			model.dataChanged.on(e => {
+				if (e.changes.hasOwnProperty('rows')) {
+					const deep = this.generateMode === 'deep' || isUndefined(this.generateMode);
```
Comment (klumba12) [inline src/ng/components/column/column.list.js]: use switch it more readable

File: src/core/column-list/generate.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+						});
+						return accessor[prop].join('; ');
+					}
+				});
+			} else {
+				const mergeColumns = generate([source[prop]], true, key);
+				acc.push(...mergeColumns);
+			}
+		} else {
+			const parts = path !== '' ? path.split('.') : [];
```
Comment (klumba12) [inline src/core/column-list/generate.js]: make a path service
const getSet|value  = path.compile(path);

File: src/core/column-list/generate.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	const props = Object.getOwnPropertyNames(source);
+	return props.reduce((acc, prop) => {
+		const key = path !== '' ?
+			`${path}.${prop}` :
+			prop;
+
+		if (deep && isObject(source[prop])) {
+			if (isArray(source[prop])) {
+				const parts = path !== '' ? path.split('.') : [];
+				acc.push({
```
Comment (klumba12) [inline src/core/column-list/generate.js]: instead of acc use memo

File: src/core/column-list/generate.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+export function generate(rows, deep=true, path='') {
+	if (!rows || rows.length === 0) {
+		return [];
+	}
+
+	const source = rows[0];
+	const props = Object.getOwnPropertyNames(source);
+	return props.reduce((acc, prop) => {
+		const key = path !== '' ?
```
Comment (klumba12) [inline src/core/column-list/generate.js]: put expression to one line

File: docs/column.generation.md [hunk 6]
```diff
@@ -1,12 +1,74 @@
...
+}
+```
+
+Column generation has two modes:
+- flat: get first-level keys of passed rows and use it for columns
+- deep: more complex mode which will handle nested objects untill doesn't find primitive type or array; it'll create getter/setter using handled path to column as well
+
+> Note that array is converted to string joined by semicolon
+
+User should set attribute generate to true for columns and generate-mode to deep|flat (deep by default) to switch on column autogeneration.
```
Comment (vkorolev) [inline docs/column.generation.md]: Maybe `first-level` or `shallow` instead of `flat`?

**Review by klumba12 — APPROVED** [review]

**Pull request merged**